Error with `contain-substring/mach-regexp` matcher and `json/junit` formatters

**Describe the bug**

I try to use the `contain-substring` matcher with the `junit`, `json` formatters but I have the following error: `panic: interface conversion: interface {} is nil, not float64`.

**How To Reproduce**

```yaml
# goss/datadog.yml
package:
  datadog-agent:
     installed: true
     versions:
       and:
         - have-len: 1
         - contain-substring: "7.1"
```

And validate it with `junit` formatter : 

```console
goss -g goss/datadog.yml validate -f junit
```

ℹ️ I also had the same issues with `ip` ressource.

**Expected Behavior**

Finishing without errors like with the `documentation` or `rspecish` formatters.

**Actual Behavior**

✅ 

```console
goss -g goss/datadog.yml validate -f documentation
Package: datadog-agent: installed: matches expectation: true
Package: datadog-agent: version: matches expectation: {"and":[{"have-len":1},{"contain-substring":"7.1"}]}
```

💥 

```console
goss -g goss/datadog.yml validate -f junit
panic: interface conversion: interface {} is nil, not float64

goroutine 1 [running]:
github.com/goss-org/goss/outputs.JUnit.Output({}, {0xc86920, 0xc000012018}, 0xc0000463c0?, {{0x11ba608?, 0xc0001c0a40?, 0xc000282400?}})
        /home/travis/gopath/src/github.com/goss-org/goss/outputs/junit.go:51 +0xe8c
github.com/goss-org/goss.ValidateConfig(0xc0001c6200, 0xc000282280)
        /home/travis/gopath/src/github.com/goss-org/goss/validate.go:142 +0x42a
github.com/goss-org/goss.Validate(0xc0001c6200)
        /home/travis/gopath/src/github.com/goss-org/goss/validate.go:111 +0x85
main.main.func1(0xc00010b1e0?)
        /home/travis/gopath/src/github.com/goss-org/goss/cmd/goss/goss.go:151 +0x25
github.com/urfave/cli.HandleAction({0xaaa1a0?, 0xbddc50?}, 0x8?)
        /home/travis/gopath/pkg/mod/github.com/urfave/cli@v1.22.14/app.go:524 +0x50
github.com/urfave/cli.Command.Run({{0xb8f1c5, 0x8}, {0x0, 0x0}, {0xc000191b20, 0x1, 0x1}, {0xb95bee, 0xf}, {0x0, ...}, ...}, ...)
        /home/travis/gopath/pkg/mod/github.com/urfave/cli@v1.22.14/command.go:175 +0x67b
github.com/urfave/cli.(*App).Run(0xc000164e00, {0xc0000341e0, 0x6, 0x6})
        /home/travis/gopath/pkg/mod/github.com/urfave/cli@v1.22.14/app.go:277 +0xb87
main.main()
        /home/travis/gopath/src/github.com/goss-org/goss/cmd/goss/goss.go:408 +0x2078
junit
```

**Environment:**
 - Version of goss: `v0.4.2`
 - OS/Distribution version: `Linux (Debian 11)`

I also tried with `mach-regexp` with same kind of issue:

```yaml
package:
  datadog-agent:
    installed: true
    versions:
      and:
        - have-len: 1
        - match-regexp: '^1:7\.\d+\.\d+(-\d+)?'
```

```console
goss -g goss/datadog.yml validate -f documentation
Package: datadog-agent: installed: matches expectation: true
Package: datadog-agent: version: matches expectation: {"and":[{"have-len":1},{"match-regexp":"^1:7\\.\\d+\\.\\d+(-\\d+)?"}]}


Total Duration: 0.020s
Count: 5, Failed: 0, Skipped: 0
```

```console
goss -g goss/datadog.yml validate -f junit
panic: interface conversion: interface {} is nil, not float64

goroutine 1 [running]:
github.com/goss-org/goss/outputs.JUnit.Output({}, {0xc86920, 0xc000012018}, 0xc00026a300?, {{0x11ba608?, 0xc0003105e0?, 0xc000237180?}})
        /home/travis/gopath/src/github.com/goss-org/goss/outputs/junit.go:51 +0xe8c
github.com/goss-org/goss.ValidateConfig(0xc0002a0200, 0xc000236f00)
        /home/travis/gopath/src/github.com/goss-org/goss/validate.go:142 +0x42a
github.com/goss-org/goss.Validate(0xc0002a0200)
        /home/travis/gopath/src/github.com/goss-org/goss/validate.go:111 +0x85
main.main.func1(0xc0002226e0?)
        /home/travis/gopath/src/github.com/goss-org/goss/cmd/goss/goss.go:151 +0x25
github.com/urfave/cli.HandleAction({0xaaa1a0?, 0xbddc50?}, 0x8?)
        /home/travis/gopath/pkg/mod/github.com/urfave/cli@v1.22.14/app.go:524 +0x50
github.com/urfave/cli.Command.Run({{0xb8f1c5, 0x8}, {0x0, 0x0}, {0xc0002f63b0, 0x1, 0x1}, {0xb95bee, 0xf}, {0x0, ...}, ...}, ...)
        /home/travis/gopath/pkg/mod/github.com/urfave/cli@v1.22.14/command.go:175 +0x67b
github.com/urfave/cli.(*App).Run(0xc0002421c0, {0xc000034080, 0x8, 0x8})
        /home/travis/gopath/pkg/mod/github.com/urfave/cli@v1.22.14/app.go:277 +0xb87
main.main()
        /home/travis/gopath/src/github.com/goss-org/goss/cmd/goss/goss.go:408 +0x2078
```

--- a/go.mod
+++ b/go.mod
@@ -11,7 +11,6 @@ require (
 	github.com/goss-org/GOnetstat v0.0.0-20230101144325-22be0bd9e64d
 	github.com/goss-org/go-ps v0.0.0-20230609005227-7b318e6a56e5
 	github.com/hashicorp/logutils v1.0.0
-	github.com/icza/dyno v0.0.0-20230330125955-09f820a8d9c0
 	github.com/miekg/dns v1.1.55
 	github.com/moby/sys/mountinfo v0.6.2
 	github.com/oleiade/reflections v1.0.1
@@ -24,7 +23,7 @@ require (
 	github.com/stretchr/testify v1.8.4
 	github.com/tidwall/gjson v1.15.0
 	github.com/urfave/cli v1.22.14
-	gopkg.in/yaml.v2 v2.4.0
+	gopkg.in/yaml.v3 v3.0.1
 	gotest.tools/v3 v3.0.3
 )
 
@@ -61,5 +60,4 @@ require (
 	golang.org/x/text v0.12.0 // indirect
 	golang.org/x/tools v0.11.1 // indirect
 	google.golang.org/protobuf v1.31.0 // indirect
-	gopkg.in/yaml.v3 v3.0.1 // indirect
 )

--- a/matcher_test.go
+++ b/matcher_test.go
@@ -58,6 +58,9 @@ func TestMatchers(t *testing.T) {
 				t.Fatal(err)
 			}
 			exitCode, err := Validate(cfg)
+			if err != nil {
+				t.Fatal(err)
+			}
 			actualOut := output.String()
 			actualOut = sanitizeOutput(actualOut)
 

--- a/outputs/json.go
+++ b/outputs/json.go
@@ -59,7 +59,7 @@ func (r Json) Output(w io.Writer, results <-chan []resource.TestResult,
 			m["successful"] = testResult.Result != resource.FAIL
 			m["summary-line"] = humanizeResult(testResult, false, includeRaw)
 			m["summary-line-compact"] = humanizeResult(testResult, true, includeRaw)
-			m["duration"] = int64(m["duration"].(float64))
+			m["duration"] = testResult.Duration.Nanoseconds()
 			resultsOut = append(resultsOut, m)
 			testCount++
 		}

--- a/outputs/junit.go
+++ b/outputs/junit.go
@@ -47,8 +47,7 @@ func (r JUnit) Output(w io.Writer, results <-chan []resource.TestResult,
 			if endTime.IsZero() || testResult.EndTime.After(endTime) {
 				endTime = testResult.EndTime
 			}
-			m := struct2map(testResult)
-			duration := strconv.FormatFloat(m["duration"].(float64)/1000/1000/1000, 'f', 3, 64)
+			duration := strconv.FormatFloat(testResult.Duration.Seconds(), 'f', 3, 64)
 			summary[testCount] = "<testcase name=\"" +
 				testResult.ResourceType + " " +
 				escapeString(testResult.ResourceId) + " " +

--- a/outputs/outputs.go
+++ b/outputs/outputs.go
@@ -16,7 +16,6 @@ import (
 	"github.com/fatih/color"
 	"github.com/goss-org/goss/resource"
 	"github.com/goss-org/goss/util"
-	"github.com/icza/dyno"
 	"github.com/pmezard/go-difflib/difflib"
 )
 
@@ -147,8 +146,6 @@ func maybeAddDiff(ss []string, expected, actual any, compact bool) []string {
 }
 
 func prettyPrint(i interface{}, indent bool) string {
-	// JSON doesn't like non-string keys
-	i = dyno.ConvertMapI2MapS(i)
 	buffer := &bytes.Buffer{}
 	encoder := json.NewEncoder(buffer)
 	encoder.SetEscapeHTML(false)

--- a/resource/gomega.go
+++ b/resource/gomega.go
@@ -27,7 +27,6 @@ func matcherToGomegaMatcher(matcher any) (matchers.GossMatcher, error) {
 		}
 		return matchers.ContainElements(interfaceSlice...), nil
 	}
-	matcher = sanitizeExpectedValue(matcher)
 	if matcher == nil {
 		return nil, fmt.Errorf("Syntax Error: Missing required attribute")
 	}
@@ -71,12 +70,16 @@ func matcherToGomegaMatcher(matcher any) (matchers.GossMatcher, error) {
 		}
 		return matchers.WithSafeTransform(matchers.ToString{}, matchers.ContainSubstring(v)), nil
 	case "have-len":
-		v, isFloat := value.(float64)
-		if !isFloat {
+		var v int
+		switch val := value.(type) {
+		case float64:
+			v = int(val)
+		case int:
+			v = val
+		default:
 			return nil, invalidArgSyntaxError("have-len", "numeric", value)
-
 		}
-		return matchers.HaveLen(int(v)), nil
+		return matchers.HaveLen(v), nil
 	case "have-patterns":
 		_, isArr := value.([]any)
 		if !isArr {
@@ -92,7 +95,7 @@ func matcherToGomegaMatcher(matcher any) (matchers.GossMatcher, error) {
 		return matchers.HaveKey(subMatcher), nil
 	case "contain-element":
 		switch value.(type) {
-		case map[string]any, string, float64:
+		case map[string]any, string, float64, int:
 		default:
 			return nil, invalidArgSyntaxError("contain-element", "matcher, string or numeric", value)
 
@@ -187,26 +190,6 @@ func sliceToGomega(value any, name string) ([]matchers.GossMatcher, error) {
 	return subMatchers, nil
 }
 
-// sanitizeExpectedValue normalizes the value so json and yaml are the same
-func sanitizeExpectedValue(i any) any {
-	if e, ok := i.(int); ok {
-		return float64(e)
-	}
-	if e, ok := i.(map[any]any); ok {
-		out := make(map[string]any)
-		for k, v := range e {
-			ks, ok := k.(string)
-			if !ok {
-				// We should never get here
-				panic(fmt.Sprintf("Matcher key type not string: %T\n\n", k))
-			}
-			out[ks] = sanitizeExpectedValue(v)
-		}
-		return out
-	}
-	return i
-}
-
 func invalidArgSyntaxError(name, expected string, value any) error {
-	return fmt.Errorf("Syntax Error: Invalid '%s' argument. Expected %s value, but received: %q", name, expected, value)
+	return fmt.Errorf("Syntax Error: Invalid '%s' argument. Expected %s value, but received: %T: %q", name, expected, value, value)
 }

--- a/resource/validate.go
+++ b/resource/validate.go
@@ -169,8 +169,6 @@ func ValidateGomegaValue(res ResourceRead, property string, expectedValue any, a
 		err = fmt.Errorf("Unknown method signature: %t", f)
 	}
 
-	foundValue = sanitizeExpectedValue(foundValue)
-	expectedValue = sanitizeExpectedValue(expectedValue)
 	var success bool
 	if gomegaMatcher == nil && err == nil {
 		gomegaMatcher, err = matcherToGomegaMatcher(expectedValue)

--- a/store.go
+++ b/store.go
@@ -10,7 +10,7 @@ import (
 	"sort"
 	"strings"
 
-	"gopkg.in/yaml.v2"
+	"gopkg.in/yaml.v3"
 
 	"github.com/goss-org/goss/resource"
 	"github.com/goss-org/goss/util"

--- a/store_test.go
+++ b/store_test.go
@@ -102,9 +102,9 @@ func Test_varsFromString(t *testing.T) {
 			name: "YAML_object",
 			arg:  `{o: {oa: a, oo: { oo1: 1 } } }`,
 			want: map[string]any{
-				"o": map[any]any{
+				"o": map[string]any{
 					"oa": "a",
-					"oo": map[any]any{
+					"oo": map[string]any{
 						"oo1": 1,
 					},
 				},

